Text Entity Improvements

HifiExperiments opened the pull request:
Closes #583 and #126 

incredibly this is NOT a protocol change 😄 

This adds support for smarter text clipping at the edges of text entities, as well as replacing the old text font format with https://github.com/Chlumsky/artery-font-format

New fonts can be created trivially from .ttf and .otf font files using https://github.com/Chlumsky/msdf-atlas-gen/ (windows download available: https://github.com/Chlumsky/msdf-atlas-gen/releases).  Eventually it would maybe be nice if we packaged the generator with our releases.

Here's the command I used to generate the default fonts (you *must* use mtsdf as the type): `msdf-atlas-gen.exe -font Roboto-Regular.ttf -type mtsdf -fontname Roboto -format png -arfont Roboto.arfont`

I think there's still some room to improve the antialiasing and other effects since right now we use a hard cutoff for the opacity, which somewhat defeats the purpose of sdfs, but that's a bit more complicated so I left it for a follow up.

## Funding

This project is funded through [NGI0 Entrust](https://nlnet.nl/entrust), a fund established by [NLnet](https://nlnet.nl) with financial support from the European Commission's [Next Generation Internet](https://ngi.eu) program. Learn more at the [NLnet project page](https://nlnet.nl/project/Overte).

[<img src="https://nlnet.nl/logo/banner.png" alt="NLnet foundation logo" width="20%" />](https://nlnet.nl)
[<img src="https://nlnet.nl/image/logos/NGI0_tag.svg" alt="NGI Zero Logo" width="20%" />](https://nlnet.nl/entrust)

HifiExperiments commented on cmake/ports/artery-font-format/disable-checksum.patch:
```diff
@@ -0,0 +1,38 @@
+-        if (footer.checksum != finalChecksum)
+-            return false;
++        //if (footer.checksum != finalChecksum)
++        //    return false;
```
this is a horrible hack, but this checksum check always failed for me with newly generated fonts, and even failed intermittently for the example font provided in https://github.com/Chlumsky/artery-font-format

a little more info here: https://github.com/Chlumsky/artery-font-format/issues/6

HifiExperiments commented:
some compiler complaints:
```
 /__w/overte/overte/libraries/procedural/src/procedural/ProceduralMaterialCache.h: In copy constructor 'NetworkMaterial::NetworkMaterial(const NetworkMaterial&)':
/__w/overte/overte/libraries/procedural/src/procedural/ProceduralMaterialCache.h:81:10: warning: 'NetworkMaterial::_lightmapParams' will be initialized after [-Wreorder]
   81 |     vec2 _lightmapParams;
      |          ^~~~~~~~~~~~~~~
/__w/overte/overte/libraries/procedural/src/procedural/ProceduralMaterialCache.h:74:10: warning:   'bool NetworkMaterial::_isOriginal' [-Wreorder]
   74 |     bool _isOriginal{ true };
      |          ^~~~~~~~~~~
/__w/overte/overte/libraries/procedural/src/procedural/ProceduralMaterialCache.cpp:825:1: warning:   when initialized here [-Wreorder]
  825 | NetworkMaterial::NetworkMaterial(const NetworkMaterial& m) :
      | ^~~~~~~~~~~~~~~
/__w/overte/overte/libraries/procedural/src/procedural/ProceduralMaterialCache.cpp: In copy constructor 'NetworkMToonMaterial::NetworkMToonMaterial(const NetworkMToonMaterial&)':
/__w/overte/overte/libraries/procedural/src/procedural/ProceduralMaterialCache.cpp:1211:31: error: call of overloaded 'NetworkMaterial(const NetworkMToonMaterial&)' is ambiguous
 1211 |     _outline(material._outline)
      |                               ^
In file included from /__w/overte/overte/libraries/procedural/src/procedural/ProceduralMaterialCache.cpp:9:
/__w/overte/overte/libraries/procedural/src/procedural/ProceduralMaterialCache.h:29:5: note: candidate: 'NetworkMaterial::NetworkMaterial(graphics::Material)'
   29 |     NetworkMaterial(const graphics::Material material) : graphics::Material(material) {}
      |     ^~~~~~~~~~~~~~~
/__w/overte/overte/libraries/procedural/src/procedural/ProceduralMaterialCache.cpp:825:1: note: candidate: 'NetworkMaterial::NetworkMaterial(const NetworkMaterial&)'
  825 | NetworkMaterial::NetworkMaterial(const NetworkMaterial& m) :
      | ^~~~~~~~~~~~~~~
```

vegaslon commented:
This text entity is not rendering correctly anymore in the tutorial https://i.gyazo.com/1118d6635fa19506c425912a96a1d927.jpg

SilverfishVR commented:
Yea, the portal text entity needs to be a tiny bit wider.

But finally, we can add fonts yay!🎆 🎉 
![Screenshot 2024-04-19 00 32 17](https://github.com/overte-org/overte/assets/53531160/4cc3eeb7-e259-43d0-afef-9fe67d257670)
It takes some fiddling with the settings to make it look nice, the default size is  way too crusty looking and the 7-segment font still has a start offset that I can't figure out how to fix. 

also, only png format works, so that should be noted in documentation too.

HifiExperiments commented:
the text is supposed to no longer completely disappear when it reaches the edge of the entity, so it sounds like there is a bug there.  I will look into that.

bmp format is also supposed to work, I will test it again

could you share the font that has the problematic start offset?

SilverfishVR commented:
bmp does not generate at all so I did not test:
`Atlas type not compatible with image format. MTSDF requires a format with alpha channel.`
png is probably preferred anyway so not really an issue.

As for the offset issue, I am an idiot 🤦 It is a fixed width font (as it should be) so if the first character is a 1, it will be offset from the edge because that is literally how 7 segment displays work.. 

Here is the 7-segment font anyway, it's SIL OFL license:
https://silverfish-freestuff.s3.amazonaws.com/TestStuff/7segment.ttf

The only other issue I found is that sometimes, but not always,  when changing to a font with a different glyph size, the text renders all scrambled at first. changing the entity size fixes it instantly:
![Screenshot 2024-04-19 22 26 37](https://github.com/overte-org/overte/assets/53531160/023834aa-7557-42bc-ae1b-ec316b061103)

ksuprynowicz reviewed (approved):
This is seriously cool :)
I reviewed it and everything looks ok

HifiExperiments commented:
@SilverfishVR I've fixed the issue in the tutorial (I was calculating the space width wrong, so the text itself was slightly wider than it should have been), updated the docs to reflect that only PNG is supported, and pushed a possible fix for the text scrambling

SilverfishVR commented:
Looks good.
I did not see the issue with text scrambling and the tutorial text is good now 👍

ksuprynowicz commented:
Thanks for testing!
Maybe we can merge it?

HifiExperiments commented:
ready to merge from my end!  I have some ideas for a follow up for how to get crisper text but those changes will be minor in the shader and I’d like to get this in first.

thanks all!
